src/type_icons.c

```diff
@@ -0,0 +1,531 @@
+{
+    .tileTag = TYPE_ICON_TAG_2,
+    .paletteTag = TYPE_ICON_TAG_2,
+    .oam = &sOamData_TypeIcons,
+    .anims = sSpriteAnimTable_TypeIcons,
+    .images = NULL,
+    .affineAnims = gDummySpriteAffineAnimTable,
+    .callback = SpriteCB_TypeIcon
+};
+
+void LoadTypeIcons(u32 battler)
+{
+    u32 position;
+
+    if (B_SHOW_TYPES == SHOW_TYPES_NEVER)
+        return;
+
+    LoadTypeSpritesAndPalettes();
+
+    for (position = 0; position < gBattlersCount; ++position)
+        LoadTypeIconsPerBattler(battler, position);
+}
+
+static void LoadTypeSpritesAndPalettes(void)
+{
+    if (IndexOfSpritePaletteTag(TYPE_ICON_TAG) != UCHAR_MAX)
+        return;
+
+    LoadCompressedSpriteSheet(&sSpriteSheet_TypeIcons1);
+    LoadCompressedSpriteSheet(&sSpriteSheet_TypeIcons2);
+    LoadCompressedSpritePalette(&sTypeIconPal1);
+    LoadCompressedSpritePalette(&sTypeIconPal2);
+}
+
+static void LoadTypeIconsPerBattler(u32 battler, u32 position)
+{
+    u32 typeNum, types[2];
+    u32 battlerId = GetBattlerAtPosition(position);
+    bool32 useDoubleBattleCoords = UseDoubleBattleCoords(battlerId);
+
+    if (IsBattlerFainted(battlerId))
+        return;
+
+    for (typeNum = 0; typeNum < 2; ++typeNum)
+        types[typeNum] = GetMonPublicType(battlerId, typeNum);
+
+    for (typeNum = 0; typeNum < 2; ++typeNum)
+        CreateSpriteFromType(position, useDoubleBattleCoords, types, typeNum, battler);
+}
+
+static bool32 UseDoubleBattleCoords(u32 position)
+{
+    if (!BATTLE_TYPE_IS_DOUBLE)
+        return FALSE;
+
+    if ((position == B_POSITION_PLAYER_LEFT) && (IsBattlerNull(B_POSITION_PLAYER_RIGHT)))
+        return FALSE;
+
+    if ((position == B_POSITION_OPPONENT_LEFT) && (IsBattlerNull(B_POSITION_OPPONENT_RIGHT)))
+        return FALSE;
+
+    return TRUE;
+}
+
+static bool32 IsBattlerNull(u32 battlerId)
+{
+    return gBattleMons[battlerId].species== SPECIES_NONE;
+}
+
+static bool32 IsBattlerFainted(u32 battlerId)
+{
+    return gBattleMons[battlerId].hp < 1;
+}
+
+static u32 GetMonPublicType(u32 battlerId, u32 typeNum)
+{
+    struct Pokemon* mon = GetBattlerData(battlerId);
+    struct Pokemon* monIllusion = GetIllusionMonPtr(battlerId);
+    u32 illusionSpecies = GetMonData(monIllusion,MON_DATA_SPECIES,NULL);
+    u32 monSpecies = GetMonData(mon,MON_DATA_SPECIES,NULL);
+
+    if (ShouldHideUncaughtType(monSpecies))
+        return TYPE_MYSTERY;
+
+    if (GetActiveGimmick(battlerId) == GIMMICK_TERA)
+        return ReturnMonDefensiveTeraType(mon,monIllusion,battlerId,typeNum,illusionSpecies,monSpecies);
+
+    if (IsIllusionActiveAndTypeUnchanged(monIllusion,monSpecies, battlerId))
+        return gSpeciesInfo[illusionSpecies].types[typeNum];
+
+	return gBattleMons[battlerId].types[typeNum];
+}
+
+static struct Pokemon* GetBattlerData(u32 battlerId)
+{
+    u32 index = gBattlerPartyIndexes[battlerId];
+
+    return (GetBattlerSide(battlerId) == B_SIDE_OPPONENT) ? &gEnemyParty[index] : &gPlayerParty[index];
+}
```

Comment:
This function should be moved to `include/battle.h` as an inline because it can clean up a lot of code. I don't think the clean up should happen here or at all but everything new should take advantage of it.

Reply:
https://github.com/rh-hideout/pokeemerald-expansion/pull/5131/commits/997fa9e9cab75dc4231694fc68a44434edf1cd96